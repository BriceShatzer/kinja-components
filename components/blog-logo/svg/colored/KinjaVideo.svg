<svg xmlns="http://www.w3.org/2000/svg" width="243" height="36" aria-label="Combined Shape" viewBox="0 0 243 36">
	<path d="M12.2.8h22.1l-5 9.3c-.7 1.1-2.4 2.6-5.7 2.7-3.3 0-11.3-.2-16.2 0-5 .1-7.2 2.8-7.2 2.8H14c2.2 0 4.2.3 5.8 2.7 2.2 3.6 9.6 17 9.6 17h-10a6 6 0 0 1-5.7-3.4L6.5 20a9 9 0 0 0-4.3-3.7c-1.3-.4-2.2-.5-2.2-.5v-.2L7 4.2C7.9 2.3 9.3.8 12.2.7zM80.4 10v11.6c0 3.2-2 4.7-4.8 4.7-1.3 0-2.4-.4-3.4-1.3l1-2.1a3 3 0 0 0 2.2 1c1.4 0 2.3-.9 2.3-2.4V10h2.7zm70.4-.3c4 0 6.4 3.6 6.4 8.3 0 4.7-2.4 8.3-6.4 8.3-4.1 0-6.5-3.6-6.5-8.3 0-4.7 2.4-8.3 6.5-8.3zm-48.4.3l3.5 12.8 3.5-12.8h2.9l-4.8 16h-3.2l-4.8-16h2.9zm-59 0v7.7l5-7.7h3.1l-5.2 7.6 5.6 8.4h-3.1l-4.2-6.6-1.2 1.8V26h-2.6V10h2.6zm12.9 0v16h-2.7V10h2.7zm130.5 0c2.7 0 4.1 1.8 4.1 4 0 2-1 3.4-2.5 3.7 1.6.3 2.8 2 2.8 4 0 2.5-1.4 4.3-4.2 4.3h-6.3V10zM62 10l5.8 10.5V10h2.7v16H68l-6-10.8V26h-2.7V10H62zm163 0c2.7 0 4.7 2 4.7 5 0 2.7-1.6 4.1-3 4.5l3 6.5h-2.9l-2.7-6h-2v6h-2.7V10zm-12 0l4.8 16H215l-.9-3.2h-5.3L208 26h-2.8l4.8-16h3.1zM89.5 10l4.8 16h-2.9l-.8-3.2h-5.3l-1 3.2h-2.8l4.8-16h3.2zm109.7 0c2.8 0 4.7 2 4.7 5 0 2.7-1.5 4.1-3 4.5L204 26h-3l-2.7-6h-2v6h-2.6V10zm-21.5 0v16H175V10h2.7zm-61.2 0v16h-2.6V10h2.6zm50.8 0v13.7h5.6V26h-8.3V10h2.7zm-43 0c4.4 0 6.8 3.4 6.8 8 0 4.7-2.4 8-6.7 8h-4.8V10zm18 0v2.3h-6v4.4h5.9V19h-6v4.7h6.2V26h-8.8V10h8.8zm91 0l3.4 6.8 3.4-6.8h2.9l-5 9.3V26h-2.6v-6.7l-5-9.3h3zm-82.5 2c-2.4 0-3.8 2.6-3.8 6 0 3.3 1.4 6 3.8 6s3.7-2.7 3.7-6c0-3.4-1.3-6-3.7-6zm35.5 7h-3v4.7h3c1.4 0 2.2-.9 2.2-2.4 0-1.3-.7-2.3-2.2-2.3zm-62-6.7h-2v11.4h2c2.7 0 4.1-2.4 4.1-5.7 0-3.4-1.4-5.7-4-5.7zm-36.4.5l-2.1 7.7H90l-2-7.7zm123.6 0l-2.2 7.7h4.3l-2.1-7.7zm-12.7-.5h-2.5v5.3h2.5c1.5 0 2.4-1 2.4-2.6s-1-2.7-2.4-2.7zm25.8 0H222v5.3h2.5c1.4 0 2.3-1 2.3-2.6s-.9-2.7-2.3-2.7zm-38.4 0h-2.8v4.4h2.8c1.4 0 2-1 2-2.2 0-1.2-.6-2.2-2-2.2z"/>
</svg>
