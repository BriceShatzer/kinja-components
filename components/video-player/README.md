# Video player

<!-- STORY -->

## Video player props
| Name | Type | Required | Default value | Description
|------|------|----------|---------------|------------
| muted | `boolean` | | | start video muted
| autoplay | `string|boolean` | | | autoplay settings
| controls | `boolean` | | | show or hide video controls
| width | `number` | ✓ | | video width
| height | `number` | ✓ | | video height
| poster | `string` | | | video poster image, thumbnail
| sources | `Array<VideoSource>` | ✓ | | video sources, documentation TBD
| adTagUrl | `string` | | | ad tag url, documentation TBD

For detailed description of each option of video.js, please see https://docs.videojs.com/tutorial-options.html
