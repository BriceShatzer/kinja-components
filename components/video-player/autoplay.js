const autoplays = new Map();

export const autoplay = playerInstance => {
	const playAttempt = playerInstance.play();

	if (!autoplays.has(playerInstance)) {
		autoplays.set(
			playerInstance,
			new Promise(resolve =>
				playAttempt
					.then(() => resolve(true))
					.catch(() => resolve(false))
			)
		);
	}

	return playAttempt;
};

export const isAutoplay = playerInstance =>
	autoplays.get(playerInstance);
