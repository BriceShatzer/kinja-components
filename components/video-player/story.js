// @flow

import * as React from 'react';
import styled from 'styled-components';
import {
	storiesOf,
	action
} from 'base-storybook';

import VideoPlayer from './video-player';

// $FlowFixMe
import 'video.js/dist/video-js.css';
// $FlowFixMe
import 'videojs-ima/dist/videojs.ima.css';

const Wrapper = styled.div`
	position: relative;
	padding-bottom: 56.25%;
`;

const videoJsOptions = {
	autoplay: false,
	controls: true,
	aspectRatio: '16:9',
	poster: 'https://blog.urbansurfer.co.uk/wp-content/uploads/2017/09/kids-skateboarders.jpg',
	sources: [
		{
			src: 'https://bitdash-a.akamaihd.net/content/MI201109210084_1/m3u8s/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.m3u8',
			type: 'application/x-mpegURL'
		}
	],
	onLoad: action('Video player loaded')
};

storiesOf('3. Elements|Custom Video', module)
	.add('Player', () => {
		return (
			<Wrapper>
				<VideoPlayer {...videoJsOptions}/>
			</Wrapper>
		);
	})
	.add('Player autoplaying', () => {
		return (
			<Wrapper>
				<VideoPlayer {...videoJsOptions} autoplay={true} unmuteAfterPreroll={true} />
			</Wrapper>
		);
	})
	.add('Player autoplaying muted', () => {
		return (
			<Wrapper>
				<VideoPlayer {...videoJsOptions} autoplay={true} />
			</Wrapper>
		);
	})
	.add('Player with pre-roll', () => {
		const adTagUrl = 'https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/single_ad_samples&ciu_szs=300x250&impl=s&gdfp_req=1&' +
			'env=vp&output=vast&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ct%3Dskippablelinear&correlator=';
		return (
			<Wrapper>
				<VideoPlayer {...videoJsOptions} adTagUrl={ adTagUrl } />
			</Wrapper>
		);
	})
	.add('Player with captions', () => {
		// https://kinja.com/api/videoupload/video/111002
		const options = {
			...videoJsOptions,
			captions: [{
				id: '11793',
				language: 'en',
				label: 'en',
				format: 'vtt',
				url: 'https://kinja.com/api/videoupload/caption/11793.vtt'
			}],
			// overrides
			poster: 'https://assets-jpcust.jwpsrv.com/thumbnails/w41xet41-720.jpg',
			sources: [{
				src: 'https://kinja-vh.akamaihd.net/i/prod/111002/111002_,180p,270p,.mp4.csmil/master.m3u8',
				type: 'application/x-mpegURL'
			}]
		};
		return (
			<Wrapper>
				<VideoPlayer {...options} />
			</Wrapper>
		);
	});
