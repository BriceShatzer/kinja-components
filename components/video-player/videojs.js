// @flow

import videojs from 'video.js';
// $FlowFixMe
import 'videojs-contrib-ads';
// $FlowFixMe
import 'videojs-ima';
// $FlowFixMe
import 'videojs-contrib-quality-levels';

export default videojs;

export type VideojsMetadata = {
	id: string,
	title: string,
	network: string,
	playingIndex?: number,
	autoplay: boolean
}
