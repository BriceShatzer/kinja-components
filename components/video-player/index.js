// @flow
import VideoPlayer, { PlayIconContainer, PlayIconSVG } from './video-player';
import { autoplay, isAutoplay } from './autoplay';
import videoAnalytics from './video-analytics';

export {
	VideoPlayer,
	autoplay,
	isAutoplay,
	videoAnalytics,
	PlayIconContainer,
	PlayIconSVG
};

export type Experiments = {|
	experimentId?: string,
	experimentVar?: null | number
|};
