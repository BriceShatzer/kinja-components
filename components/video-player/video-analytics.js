// @flow

import feature from 'bulbs-public-ads-manager/src/helpers/Feature';
import { isAutoplay } from './autoplay';

import type Videojs, { VideojsMetadata } from './videojs';
import type { Experiments } from './';

export type Config = {
	contentType: string,
	context?: { ga?: () => void }, // Window
	embedLocation?: string,
	embedTitle?: string,
	position: string,
	postPosition?: string,
	siteSection: string,
	website: string,
	experiments: Experiments
}

type StaticParams = {|
	network?: string,
	contentSource: string,
	contentType: string,
	durationBucket?: string,
	embedId: string,
	embedLocation: string,
	experimentId?: ?string,
	experimentVar?: ?number,
	player: string,
	playType: ?string,
	position: string,
	postPosition: string,
	siteSection: string,
	title?: string,
	type: string,
	videoId?: string,
	website: string,
|}

type CustomParams = {|
	errorMessage?: string,
	playType?: ?string,
	nonInteraction?: boolean
|}

type Params = { ...StaticParams, ...CustomParams, timeWatched: string };

type DimensionMap = {
	[string]: $Keys<Params>
}

const accountId = 'UA-142218-33';

const dimensionMap: DimensionMap = {
	dimension16: 'contentType',
	dimension22: 'errorMessage',
	dimension39: 'experimentId',
	dimension40: 'experimentVar',
	dimension41: 'type',
	dimension42: 'videoId',
	dimension44: 'durationBucket',
	dimension45: 'title',
	dimension50: 'player',
	dimension62: 'embedId',
	dimension63: 'embedLocation',
	dimension65: 'timeWatched',
	dimension83: 'contentType',
	dimension89: 'errorMessage',
	dimension101: 'website',
	dimension105: 'siteSection',
	dimension111: 'postPosition',
	dimension120: 'network',
	dimension121: 'position',
	dimension122: 'playType',
	dimension129: 'contentSource'
};

class VideojsAnalytics {
	autoplay: boolean;
	config: Config;
	eventLabel: string;
	staticParams: StaticParams;
	playerInstance: Videojs;
	videoMetadata: VideojsMetadata;
	timeListeners: { [string]: () => void }
	trackerName: string;
	trackerPrefix: string;
	initialLoad = true;

	constructor(playerInstance: Videojs, config: Config) {
		this.playerInstance = playerInstance;
		this.config = config;

		this.setInitialParams();
		this.setupEvents();
	}

	setInitialParams() {
		const {
			contentType,
			embedLocation,
			experiments,
			position,
			postPosition,
			siteSection,
			website
		} = this.config;

		this.staticParams = {
			contentSource: 'Kinja',
			contentType,
			embedId: 'none',
			embedLocation: embedLocation || `${window.location.hostname}${window.location.pathname}`,
			// Previously this was calling a getter in the video player component,
			// but importing it into mantle code broke amp videos.
			// Let's revert to that after we switch to magma-served amp!
			player: '7.6.0',
			playType: null,
			position,
			postPosition: String(postPosition || 0),
			siteSection,
			type: 'clip',
			website,
			...experiments
		};

		this.setupExperiments();
	}

	getDurationBucket() {
		const bucketBottom = Math.floor((this.playerInstance.duration() - 1) / 5);
		const bucketTop = bucketBottom + 1;
		return `${(bucketBottom * 5) + 1}-${bucketTop * 5}`;
	}

	createTracker() {
		if (!this.trackerName) {
			const trackerName = `kinjavideo${this.videoMetadata.id}`;

			this.trackerName = trackerName;
			this.trackerPrefix = trackerName + '.';

			window.ga('create', accountId, 'auto', trackerName);
		}
	}

	createEventLabel() {
		this.eventLabel = `kinjavideo-${this.videoMetadata.id}`;
	}

	setupExperiments() {
		const { experimentId, experimentVar } = this.config.experiments || {};

		if (experimentId && experimentVar !== null && experimentVar !== undefined && feature.isOn('enable_ga_exps')) {
			window.ga(`${this.trackerPrefix}set`, 'exp', `${experimentId}.${experimentVar}`);
		}
	}

	setupEvents() {
		this.initTimeListeners();

		// We trigger a manual event with video metadata because playerInstance.getMedia()
		// doesn't include video details usch as id or title.
		this.playerInstance.on('loadedmanualmeta', this.onManualMeta);
		// We still need to listen to 'loadedmetadata' to get the video duration, though.
		// The handler gets the duration, collects the previously stored metadata and
		// sends the 'Video Content Initiation' event when the video starts playing.
		this.playerInstance.on('loadedmetadata', this.onMeta);

		this.playerInstance.on('timeupdate', this.onTime);
		this.playerInstance.on('ads-ad-started', this.onAdImpression);
		this.playerInstance.on('adserror', this.onAdError);
		this.playerInstance.on('error', this.onContentError);
	}

	initTimeListeners() {
		this.timeListeners = {};

		// pseudo-start event, after 0.1s
		// this is needed because the video is started and paused before the ad is inserted
		// so the Video Content Start would be triggered before the Preroll Ad Start event
		const pseudoStartTime = 0.1;
		const pseudoStartName = 'Video Content Start';
		this.timeListeners[pseudoStartName] = () => {
			if (pseudoStartTime <= this.playerInstance.currentTime()) {
				this.sendEvent(pseudoStartName);
				delete this.timeListeners[pseudoStartName];
			}
		};

		// second checkers
		[3, 30].forEach(seconds => {
			const name = `${seconds}s`;
			this.timeListeners[name] = () => {
				if (seconds <= this.playerInstance.currentTime()) {
					this.sendEvent(`Video Content ${name}`);
					delete this.timeListeners[name];
				}
			};
		});
		// percentage checkers
		[25, 50, 75, 95, 100].forEach(percentage => {
			const name = `${percentage}%`;
			this.timeListeners[name] = () => {
				if (percentage <= this.playerInstance.currentTime() / this.playerInstance.duration() * 100) {
					this.sendEvent(`Video Content ${name}`);
					delete this.timeListeners[name];
				}
			};
		});
	}

	onPlayerLoad = () => {
		this.sendEvent('Video Load', { nonInteraction: true });
	};

	onManualMeta = (ev, videoMetadata) => {
		this.videoMetadata = videoMetadata;

		if (this.initialLoad) {
			this.createTracker();
		}

		this.createEventLabel();
	}

	onMeta = async () => {
		const { network, title, id } = this.videoMetadata;
		const { embedTitle } = this.config;

		this.staticParams = {
			...this.staticParams,
			network,
			title: title || embedTitle,
			videoId: id,
			durationBucket: this.getDurationBucket()
		};

		// Video Load event - should trigger once per player initiation.
		if (this.initialLoad) {
			this.onPlayerLoad();
			this.initialLoad = false;
		}

		// Now that we have all the metadata needed for our dimensions,
		// we can trigger "Video Content Initiation", but only if/when the player starts playing.
		if (!this.playerInstance.paused()) {
			this.onVideoInitiation();
		} else {
			this.playerInstance.one('play', this.onVideoInitiation);
		}

		this.initTimeListeners();
	}

	onVideoInitiation = () => {
		this.sendEvent('Video Content Initiation');
	}

	onTime = () =>
		Object.keys(this.timeListeners).forEach(key => this.timeListeners[key]());

	onAdImpression = () =>
		this.sendEvent('Preroll Ad Start 0', { nonInteraction: true });

	onAdError = ({ AdError }: { AdError: string }) =>
		this.sendEvent('Video Ad Error', {
			errorMessage: AdError,
			nonInteraction: true
		});

	onContentError = () =>
		this.sendEvent('Video Content Error', {
			errorMessage: this.playerInstance.error(),
			nonInteraction: true
		});

	async sendEvent(action: string, customParams?: CustomParams) {
		window.ga(`${this.trackerPrefix}send`, 'event', 'Video', action, this.eventLabel, await this.getDimensions(customParams));
	}

	async getDimensions(customParams?: CustomParams): {[string]: mixed} {
		const { autoplay, playingIndex } = this.videoMetadata;

		// isAutoplay tells us if an autoplay started with the autoplay() helper was successful
		const autoplayAttempt = await isAutoplay(this.playerInstance);
		// If a video is marked as autoplay but is not present in isAutoplay(), it means
		// playback hasn't begun yet so we can't know if autoplay was successful
		const isAutoplayPending = autoplay && typeof autoplayAttempt === 'undefined';

		if (!isAutoplayPending) {
			this.autoplay = autoplay && autoplayAttempt;
			this.staticParams.playType = `${(playingIndex && playingIndex > 0) ? 'Related ' : ''}${this.autoplay ? 'Autoplay' : 'Click to Play'}`;
		}

		const params: Params = {
			...this.staticParams,
			...customParams,
			timeWatched: String(Math.floor(this.playerInstance.currentTime()))
		};

		const dimensions = {};
		Object.keys(dimensionMap).forEach(dimension => {
			dimensions[dimension] = this.getValueOrDefault(params[dimensionMap[dimension]]);
		});

		// if nonInteraction is explicitly set or autoplaying
		if (customParams && customParams.nonInteraction === true || (this.autoplay && (!customParams || customParams.nonInteraction === undefined))) {
			dimensions.nonInteraction = true;
		}

		// set custom title & location fields so the event is triggered on the correct page in infinite scroll posts
		if (this.config.embedTitle && this.config.embedLocation) {
			dimensions.title = this.config.embedTitle;
			dimensions.location = this.config.embedLocation;
		}

		return dimensions;
	}

	getValueOrDefault(value: ?mixed): string {
		return (value === null || value === undefined) ? 'none' : String(value);
	}
}

const loadGoogleAnalyticsLibrary = (context = window) => {
	if (typeof window.ga === 'function') {
		// GA already loaded
		return;
	} else if (context !== window && typeof context.ga === 'function') {
		// GA is loaded in the parent document
		window.ga = context.ga;
	} else {
		// load GA
		window.ga = function () {
			window.ga.q = window.ga.q || [];
			window.ga.q.push(arguments);
		};
		const el = Object.assign(document.createElement('script'), {
			async: true,
			src: 'https://www.google-analytics.com/analytics.js'
		});

		if (document.body) {
			document.body.appendChild(el);
		}
	}
};

export default (playerInstance: Videojs, config?: Config) => {
	if (config) {
		loadGoogleAnalyticsLibrary(config.context);
		new VideojsAnalytics(playerInstance, config);
	}
};
