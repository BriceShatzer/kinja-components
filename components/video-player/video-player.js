// @flow

import * as React from 'react';
// $FlowFixMe
import hlsQualitySelector from 'videojs-hls-quality-selector';
import createScriptTag from 'kinja-components/utils/createScriptTag';
import styled from 'styled-components';
import { EnsureDefaultTheme } from '../theme';
import type Videojs from './videojs';
import type { Caption } from 'kinja-magma/api/video';
import OnScreenHOC from 'kinja-components/components/hoc/on-screen';
import { autoplay } from './autoplay';
import Logger from 'kinja-magma/utils/logger';

let logger;

// Hard-coded Play SVG icon because we want to use it as a background image in a responsive container;
// attempts to import it with raw-loader broke storybook


export const PlayIconSVG = `
	<div class="icon-wrapper">
		<svg class="icon" xmlns="http://www.w3.org/2000/svg" width="18" height="18" aria-label="PlayFilled icon" viewBox="0 0 18 18">
			<path fill="white" d="M5.3 16.7c-1.1.7-2.3 0-2.3-1.3V2.6c0-1.4 1.2-2 2.3-1.3l9.9 6.3c1 .7 1 2 0 2.7l-9.9 6.4z"/>
		</svg>
	</div>
`;

type VideoSource = {
	src: string,
	type: string
};

export type VideoPlayerProps = {|
	muted?: boolean,
	preload?: 'auto' | 'none',
	playsInline?: boolean,
	autoplay?: string | boolean,
	captions?: Array<Caption>,
	caption?: ?string,
	captionsEnabled?: boolean,
	controls?: boolean,
	width?: string,
	height?: string,
	poster: string,
	sources: Array<VideoSource>,
	adTagUrl?: ?string,
	adschedule?: ?string,
	fluid?: boolean,
	aspectRatio?: string,
	onLoad?: (playerInstance: Videojs) => void,
	isVisible?: boolean,
	playWhenInView?: boolean,
	debugMode?: boolean,
	unmuteAfterPreroll?: boolean,
	vpaidInsecureMode?: boolean
|};

type State = {
	playing?: string | boolean;
};

let version = '(unknown version)';

const Container = styled.div`
	position: relative;
	width: 100%;
	height: 100%;

	.video-js .vjs-big-play-button {
		display: none;
	}

	/* Stops caption menu from appearing on button hover, */
	/* making it only appear on click */
	.vjs-workinghover .vjs-subs-caps-button.vjs-menu-button-popup:hover {
		.vjs-menu {
			display: none;
		}
	}

	.video-js ul.vjs-menu-content {
		.vjs-menu-item {
			margin: 0;
			font-size: 0.7rem;
		}
		.vjs-menu-item-text {
			margin: 0;
			font-size: 0.7rem;
		}
	}

	video {
		width: 100%;
		height: 100%;
		cursor: pointer;
		outline: none;
	}

	.video-js {
		width: 100%;
		height: 100%;
	}
`;

const Inner = styled.div`
	.video-js {
		width: 100%;
		height: 100%;
	}
`;

export const PlayIconContainer = styled.div`
	position: absolute;
	top: 0;
	right: 0;
	left: 0;

	/* The container should not cover the control bar which is currently 30px tall at all sizes */
	bottom: ${({ withControlBar }) => withControlBar ? '30px' : '0'};

	cursor: pointer;

	.icon-wrapper {
		position: absolute;
		top: ${({ withControlBar }) => withControlBar ? 'calc(50% + 15px)' : '50%'};
		left: 50%;
		transform: translate(-50%, -50%);
		width: 15%;
		height: 0;
		padding-bottom: 15%;
		overflow: visible;
	}

	.icon {
		position: absolute;
		width: 100%;
		height: 100%;
		z-index: 22;
	}

	.icon-wrapper::after {
		content: '';
		position: absolute;
		top: -50%;
		left: -50%;
		width: 200%;
		height: 200%;
		background: radial-gradient(circle closest-side at center, rgba(0, 0, 0, 0.25), rgba(0, 0, 0, 0) 100%);
		z-index: -1;
		overflow: visible;
	}

	&:hover .icon-wrapper::after {
		background: radial-gradient(circle closest-side at center, rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0) 100%);
	}
`;

class VideoPlayer extends React.Component<VideoPlayerProps, State> {
	static get version() {
		return `VIDEOJS ${version}`;
	}

	player: Videojs;

	videoNode = React.createRef<HTMLVideoElement>();

	state: State = {
		playing: false
	};

	componentDidMount() {
		logger = new Logger('video-player', 'CLIENT_ERRORS_LOG', this.props.debugMode);

		import(/* webpackChunkName: "videojs" */ './videojs')
			.then(videojs => createScriptTag('//imasdk.googleapis.com/js/sdkloader/ima3.js')
				.then(() => this.initVideoPlayer(videojs, true))
				.catch(() => this.initVideoPlayer(videojs, false))
			);
	}

	initVideoPlayer = (videojs: Videojs, withAds: boolean) => {
		const videojsModule = typeof videojs.default === 'function' ? videojs.default : videojs;

		version = videojsModule.VERSION;

		const defaultOptions = {
			muted: true,
			controlBar: {
				pictureInPictureToggle: false
			},
			textTrackSettings: false
		};
		const options = {...defaultOptions, ...this.props};

		this.player = videojsModule(this.videoNode.current, options);

		if (this.player) {
			if (!this.player.hlsQualitySelector) {
				this.player.hlsQualitySelector = hlsQualitySelector;
			}

			this.player.hlsQualitySelector();
			withAds && this.props.adTagUrl &&
				this.player.ima({
					adTagUrl: this.props.adTagUrl,
					numRedirects: 5,
					prerollTimeout: 5000,
					vastLoadTimeout: 5000,
					debug: this.props.debugMode,
					disableCustomPlaybackForIOS10Plus: true,
					preventLateAdStart: true,
					vpaidMode: this.props.vpaidInsecureMode ? 2 : 1
				});

			this.player.on('play', () => {
				this.setState({ playing: true });
			});
			this.player.on('pause', () => this.setState({ playing: false }));
			this.player.on('loadedmetadata', () => {
				if (this.props.captions && this.props.captions.length > 0) {
					this.player.addRemoteTextTrack({
						src: this.props.captions[0].url,
						srclang: this.props.captions[0].language,
						label: 'captions on',
						kind: 'subtitles'
					}, false);
				}
			});
			this.player.on(['adserror', 'adscanceled', 'nopreroll', 'adtimeout'], e => {
				logger.error(e.type, {
					adTagUrl: this.props.adTagUrl
				});
			});
			this.player.on(['adserror', 'adscanceled', 'adend', 'adskip', 'nopreroll', 'adtimeout'], () => {
				if (this.player.paused()) {
					this.player.play();

					if (this.props.unmuteAfterPreroll) {
						// Manually unmute the player.
						// We always init the player in a state so we can play ads on iOS/Safari.
						// The next line is to make sure we have sound on the actual video once the preroll has ended.
						this.player.muted(false);
					}
				}
			});

			this.props.onLoad && this.props.onLoad(this.player);

			if (this.props.playWhenInView && this.props.isVisible) {
				autoplay(this.player);
			}
		} else {
			logger.error('Failed to initialize video.js player');
		}
	}

	pauseVideo = () => {
		if (this.state.playing) {
			this.player.pause();
			this.setState({ playing: false });
		}
	}

	componentDidUpdate(prevProps: VideoPlayerProps) {
		if (this.props.playWhenInView && !prevProps.isVisible && this.props.isVisible && this.player) {
			autoplay(this.player);
		}
	}

	// destroy player on unmount
	componentWillUnmount() {
		if (this.player) {
			this.player.dispose();
		}
	}

	render() {
		return (
			<EnsureDefaultTheme>
				<Container>
					<Inner data-vjs-player>
						<video
							className="video-js"
							// We always initialize the video as muted to make sure all browsers play our prerolls
							// - see comment in initVideoPlayer()
							muted={true}
							onTouchStart={this.pauseVideo}
							playsInline={this.props.playsInline || true}
							poster={this.props.poster}
							ref={this.videoNode}
							width="100%"
							height="100%"
						/>
					</Inner>
					{!this.state.playing && (
						<PlayIconContainer
							withControlBar
							role="button"
							tabindex="0"
							onClick={() => this.player.play()} dangerouslySetInnerHTML={{__html: PlayIconSVG}} />
					)}
				</Container>
			</EnsureDefaultTheme>
		);
	}
}

export default OnScreenHOC(VideoPlayer, {
	partialVisibility: true,
	once: true
});
