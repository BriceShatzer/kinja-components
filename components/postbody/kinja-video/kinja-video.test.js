import * as React from 'react';
import KinjaVideo from './kinja-video';
import { VideoPlayer } from '../../video-player';
import { shallow } from 'enzyme';

const props = {
	videoMeta: {
		streamingUrl: 'https://testurl.com/playlist.m3u8',
		poster: {
			id: '1234abcd',
			format: 'jpg'
		}
	}
};

describe('<KinjaVideo />', () => {
	it('should render a VideoPlayer element', () => {
		const result = shallow(<KinjaVideo {...props} />);
		expect(result.find(VideoPlayer).length).toBe(1);
		expect(result).toMatchSnapshot();
	});
});
