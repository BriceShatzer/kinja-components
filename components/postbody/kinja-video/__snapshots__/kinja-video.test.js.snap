// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KinjaVideo /> should render a VideoPlayer element 1`] = `
<kinja-video__KinjaVideoContainer
  className="js_video-wrapper"
>
  <OnScreenHOC
    aspectRatio="16:9"
    autoplay={false}
    controls={true}
    muted={true}
    playsInline={true}
    poster="https://i.kinja-img.com/gawker-media/image/upload/c_fill,f_auto,fl_progressive,g_center,h_450,q_80,w_800/1234abcd.jpg"
    preload="none"
    sources={
      Array [
        Object {
          "src": "https://testurl.com/playlist.m3u8",
          "type": "application/x-mpegURL",
        },
      ]
    }
  />
</kinja-video__KinjaVideoContainer>
`;
